apiserver/src/client/webclient.rs

```diff
@@ -0,0 +1,176 @@
+use super::error::{self, Result};
+use crate::{
+    constants::{
+        HEADER_BRUPOP_K8S_AUTH_TOKEN, HEADER_BRUPOP_NODE_NAME, HEADER_BRUPOP_NODE_UID,
+        NODE_RESOURCE_ENDPOINT,
+    },
+    CreateBottlerocketNodeRequest, DrainAndCordonBottlerocketNodeRequest,
+    UncordonBottlerocketNodeRequest, UpdateBottlerocketNodeRequest,
+};
+use models::{
+    constants::{APISERVER_SERVICE_NAME, APISERVER_SERVICE_PORT, NAMESPACE},
+    node::{BottlerocketNode, BottlerocketNodeSelector, BottlerocketNodeStatus},
+};
+
+use async_trait::async_trait;
+use snafu::ResultExt;
+use tokio::time::Duration;
+use tokio_retry::{
+    strategy::{jitter, ExponentialBackoff},
+    Retry,
+};
+use tracing::instrument;
+
+// The web client uses exponential backoff.
+// These values configure how long to delay between tries.
+const RETRY_BASE_DELAY: Duration = Duration::from_millis(100);
+const RETRY_MAX_DELAY: Duration = Duration::from_secs(10);
+const NUM_RETRIES: usize = 5;
+
+fn retry_strategy() -> impl Iterator<Item = Duration> {
+    ExponentialBackoff::from_millis(RETRY_BASE_DELAY.as_millis() as u64)
+        .max_delay(RETRY_MAX_DELAY)
+        .map(jitter)
+        .take(NUM_RETRIES)
+}
+
+#[async_trait]
+pub trait APIServerClient {
+    async fn create_bottlerocket_node(
+        &self,
+        req: CreateBottlerocketNodeRequest,
+    ) -> Result<BottlerocketNode>;
+    async fn update_bottlerocket_node(
+        &self,
+        req: UpdateBottlerocketNodeRequest,
+    ) -> Result<BottlerocketNodeStatus>;
+    async fn drain_and_cordon_node(&self, req: DrainAndCordonBottlerocketNodeRequest)
+        -> Result<()>;
+    async fn uncordon_node(&self, req: UncordonBottlerocketNodeRequest) -> Result<()>;
+}
+
+#[derive(Debug, Clone)]
+pub struct K8SAPIServerClient {
+    k8s_auth_token: String,
+}
+
+impl K8SAPIServerClient {
+    pub fn new(k8s_auth_token: String) -> Self {
+        Self { k8s_auth_token }
+    }
+
+    pub fn scheme() -> String {
+        "http".to_string()
+    }
+
+    pub fn server_domain() -> String {
+        format!(
+            "{}.{}.svc.cluster.local:{}",
+            APISERVER_SERVICE_NAME, NAMESPACE, APISERVER_SERVICE_PORT
+        )
+    }
+
+    fn add_common_request_headers(
+        &self,
+        req: reqwest::RequestBuilder,
+        node_selector: &BottlerocketNodeSelector,
+    ) -> reqwest::RequestBuilder {
+        req.header(HEADER_BRUPOP_NODE_UID, &node_selector.node_uid)
+            .header(HEADER_BRUPOP_NODE_NAME, &node_selector.node_name)
+            .header(HEADER_BRUPOP_K8S_AUTH_TOKEN, &self.k8s_auth_token)
+    }
+}
+
+#[async_trait]
+impl APIServerClient for K8SAPIServerClient {
+    #[instrument]
+    async fn create_bottlerocket_node(
+        &self,
+        req: CreateBottlerocketNodeRequest,
+    ) -> Result<BottlerocketNode> {
+        Retry::spawn(retry_strategy(), || async {
+            let http_client = reqwest::Client::new();
```

Comment:
Just curious why not use actix_web::client here to match the server.

Reply:
I actually didn't know `actix_web` had a client! I personally think that I trust `reqwest` a bit more, I know it's quite popular in the community (more than 6x more crates.io downloads it seems), but I don't have any strong reason to choose one over the other.

Reply:
I am okay with either. The reason I ask this question was I saw the actix-web-opentelemetry had the feature to support telemetry for client, but we can decide which way to go later if we want track client as well.